### ISSUE TITLE
Deploy HL to Production Environment

### ISSUE THREAD

**Issue by reddog** [healthlocker/healthlocker#1067]
Moved from #1061 to keep the flow of deployment information in a tidier place.

- [x] T6.5h - 7h for setting up the environment
  - [x] Add necessary languages/packages (t2h)
  - [x] Buffer because installing ODBC has been problematic in the past (t2h)
  - [x] Create environment variables (t25m) waiting on Adele for last 2 variables
  - [x] Write up documentation to ensure SLaM has the correct info for the server configuration (will be based off of [this documentation](https://github.com/healthlocker/oxleas-adhd/blob/master/deployment_and_ssl_doc.md) and updated where needed  to account for ePJS etc)  (T2h - 2.5h)
- [x] T2h for deployment
- [x] T25m Create database and tables on server
- [x] T2h for ssl
- [x] T25m Update Heroku DNS settings (to allow us to use healthlocker.uk)
- [x] T25m configure SSL auto-renew, see #1063
- [x] T1h for nginx configuration for the new environment
+ [ ] T1h for standups and catch ups throughout
+ [ ] T4h reserved for potential additional debugging (may not be required)

**Comment by reddog** [comment]
Production SLaM HL server is available at `51.140.71.199`

**Comment by katbow** [comment]
After call update: I will need to contact Adele or Stuart for log in details for the server. @reddog will be emailing these details to me and Rob this morning. I will also be opening an issue for the vagrantfile with details on why it's necessary to have for running the project.

**Comment by reddog** [comment]
Outbound access for SMTP and connection to MS SQL ePJS extract database not yet available (blocked by Azure firewall rules).  This has been requested.

**Comment by katbow** [comment]
@reddog can you confirm what the preferred option was for the docs? Was it to have a document on here linking to the oxleas deployment guidelines, and include steps that are different?

**Comment by katbow** [comment]
Following along the steps from oxleas deployment docs. Have run through all command up to & including installing Elixir. The erlang command was changed to `sudo apt-get install erlang-odbc` to install odbc. 

Have an error on running `mix local.hex`.
![image](https://user-images.githubusercontent.com/1287388/30424239-46144c8a-993d-11e7-99f7-065cb4cbd880.png)
Text copy:
```
 ** (FunctionClauseError) no function clause matching in Mix.Tasks.Local.run/1

    The following arguments were given to Mix.Tasks.Local.run/1:

        # 1
        ["hex"]

    (mix) lib/mix/tasks/local.ex:11: Mix.Tasks.Local.run/1
    (mix) lib/mix/task.ex:301: Mix.Task.run_task/3
    (mix) lib/mix/cli.ex:75: Mix.CLI.run_task/2
    (elixir) lib/code.ex:376: Code.require_file/2
```

**Comment by katbow** [comment]
Ahh I see. Typo in the docs which I will need to fix. Docs say `mix local hex` which I copied & pasted, but it should be `mix local.hex`.

**Comment by katbow** [comment]
For Certbot installation, will need to separate out the commands. If all of them are copied & pasted as once, it runs through the first 3 then stops. Needs to be:
```
sudo apt-get update
sudo apt-get install software-properties-common
sudo add-apt-repository ppa:certbot/certbot
```
then run
```
sudo apt-get update
sudo apt-get install python-certbot-nginx
```

**Comment by katbow** [comment]
In addition to the environment variables that need to be added for oxleas, Healthlocker also need environment variables for:
```
HEROKU_URL
SEGMENT_WRITE_KEY
READ_ONLY_HOSTNAME
READ_ONLY_USERNAME
READ_ONLY_DATABASE
READ_ONLY_PASSWORD
MPass
```

Finding all of these now & adding.

**Comment by katbow** [comment]
Have all of the above added besides `READ_ONLY_PASSWORD` and `MPass`. Will be calling Adele shortly to get these.

**Comment by reddog** [comment]
@katbow In reply to:

> can you confirm what the preferred option was for the docs? Was it to have a document on here linking to the oxleas deployment guidelines, and include steps that are different?

@iteles suggested elsewhere that a single, *master*, document be created (ie the current Oxleas document) and then differences to that document be recorded in the deployment doc for this repo.

This wouldn't be a problem *now* but rather in the *future* it could become an issue.

We need to avoid a situation where the document for this repo says something like "follow the Oxleas document to step 3, then do something else, and carry on from Oxleas step 4", and then at *some point in the future* there could be a change to the Oxleas document inserting a new step 3 (for example).  This would then **break** the document here.

@RobStallion's suggestion was to make it clear in the Oxleas doc that it is used as a base for other documents and to explicitly list those.  It would seem that this would do a great job of keeping a single (so as to help reduce maintenance of that document) document with most of the information required, plus ensuring that those making future alterations are not innocently ignorant to the impact changes would have on other documents (ie the one for this repo).

**Comment by katbow** [comment]
@reddog great, thanks! I'll list that Healthlocker is dependent on those docs in https://github.com/healthlocker/oxleas-adhd/blob/master/deployment_and_ssl_doc.md and any updates to that need to be checked alongside the docs for this repo.

**Comment by katbow** [comment]
I am running through [creating an ssl certificate](https://github.com/healthlocker/oxleas-adhd/blob/master/deployment_and_ssl_doc.md) and have run into a couple of errors. I'll need to make some changes to this section for clarity (will be opening an issue on oxleas with the changes I need to make).

The current error is after generating the cert and adding the file to the repo and pushing up, I get an error after running `mix edeliver build release`
![image](https://user-images.githubusercontent.com/1287388/30428528-d5d4882c-994b-11e7-8fbf-02bd6f29f143.png)

**Comment by reddog** [comment]
@katbow 

> Healthlocker also need environment variables for:
> `HEROKU_URL`

Are there any references to `HEROKU_URL` other than https://github.com/healthlocker/healthlocker/blob/master/config/prod.exs#L22 ?  This line is commented out (maybe should be deleted even?).

**Comment by katbow** [comment]
you're correct, I didn't realise it was commented out when I was searching through the code for if the variables were used. Thanks!

**Comment by reddog** [comment]
@katbow The instructions at https://github.com/healthlocker/oxleas-adhd/blob/master/deployment_and_ssl_doc.md require (temporary) deployment of https://github.com/RobStallion/azure_deployment_test (which needs alteration before deploying, see `.deliver/config` for example.

The error you have might be due to

`cd '/home/oxleasadmin/azure_deployment_test/builds`

`~/oxleasadmin` won't exist on this server (there's a different username).

However, https://github.com/healthlocker/healthlocker/blob/master/nginx.config#L11 has much simpler config for setting up Let's Encrypt (using version of `nginx.config` before https://github.com/healthlocker/healthlocker/blob/master/nginx.config#L9 has been added).  Copy the `well-known` file generated by Let's Encrypt to `/var/www/html` and it should be good to go without the extra interim Elixir release.

There is an even simpler method by using

```
sudo certbot --nginx certonly
```

See `advanced` panel at https://certbot.eff.org/#ubuntuxenial-nginx

> the Nginx plugin with certonly does the following:
> - make temporary config changes
> - (adding a new server block to pass an ACME Challenge)
> - performs a graceful reload
> - reverts all changes
> - performs another graceful reload

I appreciate that maybe these different routes are out of scope for this piece of work, but wanted to suggest them as options if you are encountering difficulties.

**Comment by katbow** [comment]
Got it fixed for building the release.

![image](https://user-images.githubusercontent.com/1287388/30429824-ce223afc-9950-11e7-81eb-e6d887634d14.png)

I went back to the original instructions on https://github.com/dwyl/learn-phoenix-framework/blob/master/production-deployment.md#161-install-build-tools-on-the-virtual-machine

I think there were some build tools that we may have installed earlier on the oxleas live vm, that Rob didn't need to install as he was creating the docs. I will add that to the set up section.

The commands I ran from there were:
`wget https://packages.erlang-solutions.com/erlang-solutions_1.0_all.deb && sudo dpkg -i erlang-solutions_1.0_all.deb`
`sudo apt-get update`
```
curl -sL https://deb.nodesource.com/setup_6.x | sudo -E bash -
sudo apt-get install elixir erlang-base-hipe build-essential erlang-parsetools erlang-dev nodejs -y
```

**Comment by katbow** [comment]
@reddog 

>The error you have might be due to

cd '/home/oxleasadmin/azure_deployment_test/builds

It will create a directory for whatever it's named. So even though oxleasadmin didn't exist, it was still creating the directory for it when trying to build the release. All the same I changed this for clarity so that it's set to hladmin & will delete the previously made directory. 

Thank you for those options! I will be continuing going through it as it was done by Rob since I have found the problem now, but if I encounter larger issues then I will take a look at that.

**Comment by reddog** [comment]
@katbow Thanks for that.

It seems that this interim release has flagged up missing dependencies which is good to add in the document.

The nginx plugin for certbot will remove a number of unnecessary steps - I'll add that as a suggested change to the document for future usage.

**Comment by katbow** [comment]
I called Adele but it went to voicemail. I left a message saying I was calling from dwyl for the database password & password to decrypt the user data and left my number & email for her.

**Comment by katbow** [comment]
I was carrying on work through [this section](https://github.com/healthlocker/oxleas-adhd/blob/master/deployment_and_ssl_doc.md#create-an-ssl-certificate) and realised it won't work until I move over the url so that the healthlocker.uk points to this IP! I have set that up & will carry on with that point once it's available.

**Comment by reddog** [comment]
@katbow I'll let you know if I hear anything from Adele.

Where have you made the IP address change?  Heroku and/or iwantmyname.net ?  I once came across an issue listing the original DNS configuration but I can't find that again right now.

And separately, are you having to wait for the IP address change to be available from your computer?  It can take a while for DNS changes to click over, but you can work around that by adding an entry to your local `.hosts` file.  Let me know if you need any more info on this.

**Comment by katbow** [comment]
Thanks for chasing up on that, David. If you do hear anything from her, I do also need the password for decrypting the user data. I did say this on the voicemail but in the doc you shared only the database password is mentioned to get from Adele. 

I removed the healthlocker.uk and www.healthlocker.uk from heroku, so the heroku app is now accessed with https://healthlocker.herokuapp.com/. I have updated the DNS settings on iwantmyname.

I have to wait for it for finishing the SSL cert as it needs access to `http://healthlocker.uk/...`. There are other steps I can take in the meantime, but if it gets to the point where I need to get back to that, then I will do that.

**Comment by reddog** [comment]
Can I confirm that the live site will be accessible from both `healthlocker.uk` and `www.healthlocker.uk` ?

Thanks for confirming the iwantmyname change.  Their nameservers are showing the change to the www.healthlocker.uk version but not the root healthlocker.uk entry.

With regard to the data decryption key, Garry has previously said that the test and live ePJS extract databases are populated in the same manner.  In which case, it might be just as worthwhile to try the key you already have.  This isn't the case for the database password.

**Comment by katbow** [comment]
that's right yes. It will be accessible at both `healthlocker.uk` and `www.healthlocker.uk`.

I made a mistake when I first entered updated the settings for healthlocker.uk which I noticed and now changed so it will take a bit longer than www version.

Okay great thanks for that info. I'll add that for the decryption pass.

**Comment by katbow** [comment]
Attempted to deploy healthlocker with the server set up.

Had this error:
![image](https://user-images.githubusercontent.com/1287388/30439082-c1edd1a0-996a-11e7-952e-1700de5e8e67.png)

Error text:
```
==> Release failed, during .boot generation:
        Undefined applications: [odbc]

A remote command failed on:

  root@51.140.71.199

Output of the command is shown above and the command executed
on that host is printed below for debugging purposes:

FAILED with exit status 1:

    [ -f ~/.profile ] && source ~/.profile
    set -e
    cd /home/hladmin/healthlocker/builds
    if [ "mix" = "rebar" ]; then
      echo "using rebar to generate release"
      ./rebar   -f generate
    elif [ "mix" = "relx" ]; then
      echo "using relx to generate release"
      ./relx release
    elif [ "mix" = "mix" ]; then
      echo "using mix to generate release"
      MIX_ENV="prod" LINK_SYS_CONFIG="" LINK_VM_ARGS="" APP="healthlocker" AUTO_VERSION="" BRANCH="master" SKIP_RELUP_MODIFICATIONS="" RELUP_MODIFICATION_MODULE="" USING_DISTILLERY="true" mix  release --verbose --env="prod" --name="healthlocker"
    fi
```
Double checking odbc installation for vagrant.

**Comment by katbow** [comment]
Found that quickly, forgot to install the drivers for odbc. Will be installing these & retrying the deploy.

**Comment by katbow** [comment]
I was able to deploy the app on http but it looks really weird & won't load any images or any page besides home. I was looking at the server running, and it was not even picking up on requests to other pages and was going straight to 404. 

I am not sure at this point if it is because I haven't completed the SSL or finished the nginx config, or if this is a firewall issue of some sort. Plan for checking this will be to redeploy the test app & see if anything else loads, set up the ssl and nginx config then redeploy this app. Should be able to see with all that if it is a healthlocker specific configuration issue, or if it is something with the server itself.

![image](https://user-images.githubusercontent.com/1287388/30442808-76a2da14-9975-11e7-940c-04cf7b98c5bb.png)

**Comment by reddog** [comment]
@katbow Were you accessing the site by IP address or FQDN? Could it be that the DNS changes hadn't fully propagated?

**Comment by katbow** [comment]
@dbower I tried both.

**Comment by reddog** [comment]
Comparing the HealthLocker and Oxleas `nginx.config` files at https://github.com/healthlocker/healthlocker/blob/master/nginx.config and https://github.com/healthlocker/oxleas-adhd/blob/master/nginx.config respectively the only (meaningful) changes are the two `server_name` lines.  It isn't that?

**Comment by katbow** [comment]
The nginx config isn't set up like either of those yet while doing the ssl. I had the same error on the test site. Checked the nginx config and needed to remove `try_files $uri $uri/ =404;` in the location object, which loaded the other assets.

Now have a different error after I clicked to continue for ssl
![image](https://user-images.githubusercontent.com/1287388/30476743-d784564a-9a02-11e7-8e50-0f74113be487.png)

**Comment by katbow** [comment]
Looks like it was some kind of time out issue. Ran the exact some thing again and it worked fine.

**Comment by reddog** [comment]
@katbow Good job for spotting that.  Where is the `try_files $uri $uri/ =404;` line in the docs?  Trying to keep half an eye on the deployment process and I can't find that step.

Looks like the timeout error was against `letsencrypt.org` and glad that it worked second time around.  Could have just been *one of those things*, but we'll need to be aware if we keep getting external web access timeout requests (in which case maybe there's something else at fault).

**Comment by katbow** [comment]
it's not in the docs yet. I will be adding a section for updates that need to be done to nginx specifically for getting ssl set up. I've been updating the docs as I go along. There's in-progress PRs both here & on oxleas which I have been pushing to. You'll so those updates there later today 👍 .

**Comment by katbow** [comment]
![image](https://user-images.githubusercontent.com/1287388/30478438-eecea46c-9a08-11e7-9981-fbc634f730c3.png)

Healthlocker deploy loads everything correctly now as well. Note that is still using http. Updating nginx config next to switch to https.

**Comment by reddog** [comment]
@katbow Thanks for that.  I had checked the docs and the suggest test repo at https://github.com/RobStallion/azure_deployment_test

Sounds like it would certainly be worthwhile at least trying the certbot nginx plugin next time around.  It would be a very quick check to see if that method worked.

**Comment by katbow** [comment]
https://www.healthlocker.uk/ 🎉 
![image](https://user-images.githubusercontent.com/1287388/30479299-8ee72ed0-9a0c-11e7-9919-56589e04ec0b.png)

**Comment by reddog** [comment]
( @katbow nothing wrong with a bit of self-congratulations! ;-) )

**Comment by katbow** [comment]
One last thing to do on this is update the admin roles. @katyedwards @MollieCourtenay @reddog which email addresses need to be added for this? I will add them in after lunch.

**Comment by reddog** [comment]
@katbow What is the email address of the admin user on the staging site?  Can then be used initially unless we get confirmation of something different?

**Comment by MollieCourtenay** [comment]
I think It would be good to have a new account with Healthlocker@slam.nhs.uk set up for admin. 
Could you create a new account and let us know the log details?
So that Katy's account is not admin on production.

**Comment by reddog** [comment]
@MollieCourtenay Tiny thing - could we have a lower case "H"?  Not sure if the login process does a case sensitive check against email address, but since email addresses are usually in lower case, entering with an initial capital letter might cause some frustration.

**Comment by MollieCourtenay** [comment]
Yep just auto corrects on my phone @reddog

**Comment by reddog** [comment]
(changed - thanks)

**Comment by katbow** [comment]
Emails are saved and checked as lowercase for that exact reason (auto capitalisation on phones). I'll add that email in. I'll message you the login details @MollieCourtenay  and you can pass along to anyone who needs them if that's ok.

**Comment by katbow** [comment]
The admin role has been added for that email and details have been sent to @MollieCourtenay

**Comment by katbow** [comment]
@reddog can you confirm the documentation can be merged in so that this issue can be closed? Thanks!

**Comment by reddog** [comment]
@katbow Following yesterday's testing, could you make a comment about quoting (or escaping if required) the database password to protect it from the shell and ensure it's correctly passed through to the Phoenix app?

**Comment by katyedwards** [comment]
@reddog are you happy to close this one now?

**Comment by reddog** [comment]
Other than newly discovered ePJS issues, see #1078 etc, the deployment itself is complete.
